pkg/k8s/watchers/service_test.go

```diff
@@ -2470,3 +2471,120 @@ func Test_addK8sSVCs_ExternalIPs(t *testing.T) {
 	require.EqualValues(t, del1stWanted, del1st)
 	require.EqualValues(t, del2ndWanted, del2nd)
 }
+
+func TestHeadless(t *testing.T) {
```

Comment:
Hmm that's true I am not super familiar with those tests but I can try to dig a bit more. This seems to affect some other tests in here too though:
```
❯ go test ./pkg/k8s/watchers/ --run TestChangeSVCPort --count=1
time="2024-11-11T16:03:13.74922495+01:00" level=warning msg="service not found" k8sNamespace=bar k8sSvcName=foo obj="{AddrCluster:172.0.20.1 L4Addr:{Protocol:ANY Port:80} Scope:0}" subsys=k8s-watcher
--- FAIL: TestChangeSVCPort (0.00s)
    service_test.go:483:
        	Error Trace:	/home/aouthenin/cilium/pkg/k8s/watchers/service_test.go:483
        	Error:      	Not equal:
        	            	expected: []loadbalancer.SVC{loadbalancer.SVC{Frontend:loadbalancer.L3n4AddrID{L3n4Addr:loadbalancer.L3n4Addr{AddrCluster:types.AddrCluster{addr:netip.Addr{addr:netip.uint128{hi:0x0, lo:0xffffac001401}, z:unique.Handle[net/netip.addrDetail]{value:(*netip.addrDetail)(0xc000013818)}}, clusterID:0x0}, L4Addr:loadbalancer.L4Addr{Protocol:"UDP", Port:0x50}, Scope:0x0}, ID:0x0}, Backends:[]*loadbalancer.Backend{(*loadbalancer.Backend)(0xc000a7a200)}, Type:"ClusterIP", ForwardingMode:"", ExtTrafficPolicy:"", IntTrafficPolicy:"", NatPolicy:"", SessionAffinity:false, SessionAffinityTimeoutSec:0x0, HealthCheckNodePort:0x0, Name:loadbalancer.ServiceName{Namespace:"", Name:"", Cluster:""}, LoadBalancerSourceRanges:[]*cidr.CIDR(nil), L7LBProxyPort:0x0, LoopbackHostport:false, Annotations:map[string]string(nil)}, loadbalancer.SVC{Frontend:loadbalancer.L3n4AddrID{L3n4Addr:loadbalancer.L3n4Addr{AddrCluster:types.AddrCluster{addr:netip.Addr{addr:netip.uint128{hi:0x0, lo:0xffffac001401}, z:unique.Handle[net/netip.addrDetail]{value:(*netip.addrDetail)(0xc000013818)}}, clusterID:0x0}, L4Addr:loadbalancer.L4Addr{Protocol:"UDP", Port:0x51}, Scope:0x0}, ID:0x0}, Backends:[]*loadbalancer.Backend{(*loadbalancer.Backend)(0xc000a7a280)}, Type:"ClusterIP", ForwardingMode:"", ExtTrafficPolicy:"", IntTrafficPolicy:"", NatPolicy:"", SessionAffinity:false, SessionAffinityTimeoutSec:0x0, HealthCheckNodePort:0x0, Name:loadbalancer.ServiceName{Namespace:"", Name:"", Cluster:""}, LoadBalancerSourceRanges:[]*cidr.CIDR(nil), L7LBProxyPort:0x0, LoopbackHostport:false, Annotations:map[string]string(nil)}}
        	            	actual  : []loadbalancer.SVC{loadbalancer.SVC{Frontend:loadbalancer.L3n4AddrID{L3n4Addr:loadbalancer.L3n4Addr{AddrCluster:types.AddrCluster{addr:netip.Addr{addr:netip.uint128{hi:0x0, lo:0xffffac001401}, z:unique.Handle[net/netip.addrDetail]{value:(*netip.addrDetail)(0xc000013818)}}, clusterID:0x0}, L4Addr:loadbalancer.L4Addr{Protocol:"ANY", Port:0x50}, Scope:0x0}, ID:0x0}, Backends:[]*loadbalancer.Backend{(*loadbalancer.Backend)(0xc000a7a380)}, Type:"ClusterIP", ForwardingMode:"", ExtTrafficPolicy:"", IntTrafficPolicy:"", NatPolicy:"", SessionAffinity:false, SessionAffinityTimeoutSec:0x0, HealthCheckNodePort:0x0, Name:loadbalancer.ServiceName{Namespace:"", Name:"", Cluster:""}, LoadBalancerSourceRanges:[]*cidr.CIDR(nil), L7LBProxyPort:0x0, LoopbackHostport:false, Annotations:map[string]string(nil)}, loadbalancer.SVC{Frontend:loadbalancer.L3n4AddrID{L3n4Addr:loadbalancer.L3n4Addr{AddrCluster:types.AddrCluster{addr:netip.Addr{addr:netip.uint128{hi:0x0, lo:0xffffac001401}, z:unique.Handle[net/netip.addrDetail]{value:(*netip.addrDetail)(0xc000013818)}}, clusterID:0x0}, L4Addr:loadbalancer.L4Addr{Protocol:"ANY", Port:0x51}, Scope:0x0}, ID:0x0}, Backends:[]*loadbalancer.Backend{(*loadbalancer.Backend)(0xc000a7a400)}, Type:"ClusterIP", ForwardingMode:"", ExtTrafficPolicy:"", IntTrafficPolicy:"", NatPolicy:"", SessionAffinity:false, SessionAffinityTimeoutSec:0x0, HealthCheckNodePort:0x0, Name:loadbalancer.ServiceName{Namespace:"", Name:"", Cluster:""}, LoadBalancerSourceRanges:[]*cidr.CIDR(nil), L7LBProxyPort:0x0, LoopbackHostport:false, Annotations:map[string]string(nil)}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -20,3 +20,3 @@
        	            	     L4Addr: (loadbalancer.L4Addr) {
        	            	-     Protocol: (string) (len=3) "UDP",
        	            	+     Protocol: (string) (len=3) "ANY",
        	            	      Port: (uint16) 80
        	            	@@ -51,3 +51,3 @@
        	            	      L4Addr: (loadbalancer.L4Addr) {
        	            	-      Protocol: (string) (len=3) "UDP",
        	            	+      Protocol: (string) (len=3) "ANY",
        	            	       Port: (uint16) 8080
        	            	@@ -97,3 +97,3 @@
        	            	     L4Addr: (loadbalancer.L4Addr) {
        	            	-     Protocol: (string) (len=3) "UDP",
        	            	+     Protocol: (string) (len=3) "ANY",
        	            	      Port: (uint16) 81
        	            	@@ -128,3 +128,3 @@
        	            	      L4Addr: (loadbalancer.L4Addr) {
        	            	-      Protocol: (string) (len=3) "UDP",
        	            	+      Protocol: (string) (len=3) "ANY",
        	            	       Port: (uint16) 8080
        	Test:       	TestChangeSVCPort
FAIL
FAIL	github.com/cilium/cilium/pkg/k8s/watchers	0.019s
FAIL
```

Reply:
My guess is that this is caused by a few other tests in the same file setting `option.Config.LoadBalancerProtocolDifferentiation = true` and never reverting it at the end of the test.

Reply:
Ohh yes indeed! AFAICT this option should be set to true by default so it would makes sense to force it to true as some other test also does. I did this for `TestChangeSVCPort` that was missing this too